### ISSUE TITLE
Don't issue assignment-from-none if None is returned explicitly

### ISSUE THREAD

**Issue by mthuurne** [PyCQA/pylint#2332]
### Steps to reproduce

Run Pylint on the following code:
```python
class Clazz:

    def get_value(self):
        return None

    def check_value(self):
        value = self.get_value()
        if value is not None:
            print('returned value:', value)
```

### Current behavior

The following message is issued:
```
7:8: E1128: Assigning to function call which only returns None (assignment-from-none)
```

### Expected behavior

While `Class.get_value()` indeed always returns None, this method can be overridden in a subclass to return a different value. I think it would make sense to treat a function that returns nothing different from a function that explicitly returns `None`. In other languages, the former would be a function with return type `void`, while the latter would be returning `null`. In Python `None` is used for both, but an implicit vs explicit return of `None` could be used to tell the difference (or type annotations if the program uses those).

I think it would be better if `assignment-from-none` is only issued if the called function doesn't return a value (no `return` statement or `return` without a value). If the function explicitly returns `None`, assume that `None` represents a "null" value and don't issue  `assignment-from-none`.

The message string should also be updated a bit then: instead of "Assigning to function call which only returns None" it should become something like "Using result from function call which does not return a value". (Why is assignment singled out? I think every use of the return value is equally suspect.)

### pylint --version output

```
pylint 2.0.0
astroid 2.0.0.dev4
Python 3.4.6 (default, Mar 22 2017, 12:26:13) [GCC]
```

**Comment by PCManticore** [comment]
I think the current behaviour is fine, it can be disabled locally if you have any situation where you know that a method is going to be reimplemented or not. Otherwise we'd assign a special case to `return None` which in turn might itself hide some potential errors. The linter (or this particular check) cannot predict if a method is going to be overridden or not, so it just plays on the safe side and emits the message all the time. Otherwise as mentioned it would mean assigning a weak heuristic to `return None`. In this particular case it probably would make more sense to make the method an abstract method, which is a stronger indicator that the method might be overridden later on the future, also precluding the linter for emitting the error.

Also we do currently make a distinction between not returning anything (`assignment-from-no-return`) and returning `None` (`assignment-from-none`), so you can only disable the latter if you wish.

**Comment by ssbarnea** [comment]
Should I open a bug for the example below, which triggers two rules, when in fact it should trigger none?;
```
        match = self._check_name(
            data["name"], filename=file, linenumber=data[LINE_NUMBER_KEY]
        )
        if match:
            return [match]
        return []
```

```
E1128: Assigning result of a function call, where the function returns None (assignment-from-none)
R1711: Useless return at end of function or method (useless-return)
```

**Comment by Pierre-Sassoulas** [comment]
@ssbarnea yes, I'm not sure I understood everything as I don't know what is in ``self._check_name`` so a detailed issue would help :)

**Comment by marcinsulikowski** [comment]
What about the case where code has type annotations, like the following?
```py
# pylint: disable=missing-docstring

class BaseClass():
    def get_value(self) -> int | None:
        return None

    def check_value(self) -> None:
        value = self.get_value()
        if value is not None:
            print(value)
```
Here it is pretty much obvious that the code is correct but Pylint still complains.

If Pylint is unable to understand annotations and cannot reliably tell that the annotation allows non-`None` values, maybe there could be a heuristic that `assignment-from-none` is disabled for functions with any type annotation, leaving only `assignment-from-no-return` enabled for them? That would be still sufficient to protect against bugs like this one:
```py
def do_something() -> None:
    print("something")

result = do_something()
```
This approach would cause false negatives for code like this:
```py
def do_something() -> None:
    print("something")
    return None

result = do_something()
```
However, I would worry less about false negatives in such cases because when code has type annotations, someone probably uses tools like `mypy` that implement the same check in a fully type-aware way end correctly emit `error: "do_something" does not return a value`.

If Pylint left `assignment-from-no-return` enabled for type-annotated functions, there would still be a chance for false positives but that'd be only in cases where someone omits `return None`:
```py
def get_value() -> None | int:
    print("something")

result = do_something()
```
which don't seem to be likely – I think that in most cases where something returns `Optional[X]` or `X | None` there is actually an explicit `return None` in the code, and just adding `return None` avoids the Pylint's warning.
